File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatRoom/MSChatRoomController.m
//
//  MSChatRoomController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/10/28.
//

#import "MSChatRoomController.h"
#import "MSIMSDK-UIKit.h"
#import <MSIMSDK/MSIMSDK.h>
#import <AVFoundation/AVFoundation.h>
#import "MSChatRoomController+More.h"


@interface MSChatRoomController ()<MSInputViewControllerDelegate,MSChatRoomMessageControllerDelegate>

@property(nonatomic,strong) MSChatRoomMessageController *messageController;

@property(nonatomic,strong) MSInputViewController *inputController;

@end

@implementation MSChatRoomController

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupViews];
}

- (void)dealloc
{
    MSLog(@"%@ dealloc",self.class);
}

- (void)setRoomInfo:(MSGroupInfo *)roomInfo
{
    _roomInfo = roomInfo;
    self.messageController.roomInfo = roomInfo;
}

- (void)setupViews
{
    self.messageController = [[MSChatRoomMessageController alloc]init];
    self.messageController.delegate = self;
    self.messageController.roomInfo = self.roomInfo;
    self.messageController.view.frame = CGRectMake(0, 0, Screen_Width, Screen_Height-TTextView_Height-Bottom_SafeHeight);
    [self addChildViewController:self.messageController];
    [self.view addSubview:self.messageController.view];
    
    self.inputController = [[MSInputViewController alloc]initWithChatType:MSIM_CHAT_TYPE_CHATROOM delegate:self];
    self.inputController.view.frame = CGRectMake(0, Screen_Height-TTextView_Height-Bottom_SafeHeight, Screen_Width, TTextView_Height+Bottom_SafeHeight);
    self.inputController.view.autoresizingMask = UIViewAutoresizingFlexibleTopMargin;
    [self addChildViewController:self.inputController];
    [self.view addSubview:self.inputController.view];
}

#pragma mark - <MSInputViewControllerDelegate>

- (void)inputController:(MSInputViewController *)inputController didChangeHeight:(CGFloat)height
{
    WS(weakSelf)
    [UIView animateWithDuration:0.3 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        CGRect msgFrame = weakSelf.messageController.view.frame;
        msgFrame.size.height = weakSelf.view.frame.size.height-height;
        weakSelf.messageController.view.frame = msgFrame;
        
        CGRect inputFrame = weakSelf.inputController.view.frame;
        inputFrame.origin.y = msgFrame.origin.y + msgFrame.size.height;
        inputFrame.size.height = height;
        weakSelf.inputController.view.frame = inputFrame;
        [weakSelf.messageController scrollToBottom:NO];
        } completion:^(BOOL finished) {
            
        }];
}

- (void)inputController:(MSInputViewController *)inputController didSendTextMessage:(NSString *)msg
{
    MSIMTextElem *textElem = [[MSIMManager sharedInstance] createTextMessage:msg];
    [self sendMessage:textElem];
}

- (void)inputController:(MSInputViewController *)inputController didSendVoiceMessage:(NSString *)filePath
{
    NSURL *url = [NSURL fileURLWithPath:filePath];
    AVURLAsset *audioAsset = [AVURLAsset URLAssetWithURL:url options:nil];
    NSInteger duration = (NSInteger)CMTimeGetSeconds(audioAsset.duration);
    NSInteger length = (NSInteger)[[[NSFileManager defaultManager] attributesOfItemAtPath:filePath error:nil] fileSize];
    
    MSIMVoiceElem *voiceElem = [[MSIMVoiceElem alloc]init];
    voiceElem.path = filePath;
    voiceElem.duration = duration;
    voiceElem.dataSize = length;
    voiceElem = [[MSIMManager sharedInstance] createVoiceMessage:voiceElem];
    [self sendMessage:voiceElem];
}

- (void)inputControllerDidInputAt:(MSInputViewController *)inputController
{
    
}

/**
 *  输入框中内容发生变化时的回调
 */
- (void)inputController:(MSInputViewController *)inputController contentDidChanged:(NSString *)text
{

}

/**
 *  有 @xxx 字符删除
 */
- (void)inputController:(MSInputViewController *)inputController didDeleteAt:(NSString *)atText
{
    
}

/// 点击拍照，照片等更多功能
- (void)inputController:(MSInputViewController *)inputController didSelectMoreCell:(MSInputMoreCell *)cell
{
    if ([self.delegate respondsToSelector:@selector(chatController:onSelectMoreCell:)]) {
        [self.delegate chatController:self onSelectMoreCell:cell];
    }
    if (cell.data.tye == MSIM_MORE_PHOTO) {//照片
        [self selectPhotoForSend];
    }else if (cell.data.tye == MSIM_MORE_VIDEO) {//视频
        [self selectVideoForSend];
    }else if (cell.data.tye == MSIM_MORE_LOCATION) {//位置
        [self selectLocationForSend];
    }
}

/// 点击发送自定义表情
- (void)inputController:(MSInputViewController *)inputController didSendEmotion:(BFFaceCellData *)data
{
    [self sendEnotionMessage:data];
}


#pragma mark - <MSMessageControllerDelegate>

/**
 收到信令消息
 */
- (void)messageController:(MSChatRoomMessageController *)controller onRecieveSignalMessage:(NSArray <MSIMElem *>*)elems
{
    for (MSIMElem *elem in elems) {
        if (![elem isKindOfClass:[MSIMCustomElem class]]) return;
//        MSIMCustomElem *customElem = (MSIMCustomElem *)elem;
//        NSDictionary *dic = [customElem.jsonStr el_convertToDictionary];
    }
}

/**
 *  收到新消息的函数委托
 */
- (MSMessageCellData *)messageController:(MSChatRoomMessageController *)controller prepareForMessage:(MSIMElem *)data
{
    if ([self.delegate respondsToSelector:@selector(chatController:prepareForMessage:)]) {
        return [self.delegate chatController:self prepareForMessage:data];
    }
    return nil;
}

/**
 *  显示消息数据委托
 *  您可以通过该回调实现：根据传入的 data 初始化消息气泡并进行显示
 */
- (Class)messageController:(MSChatRoomMessageController *)controller onShowMessageData:(MSMessageCellData *)data
{
    if ([self.delegate respondsToSelector:@selector(chatController:onShowMessageData:)]) {
        return [self.delegate chatController:self onShowMessageData:data];
    }
    return nil;
}

/**
 *  控制器点击回调
 *  您可以通过该回调实现：重置 InputControoler，收起键盘。
 */
- (void)didTapInMessageController:(MSChatRoomMessageController *)controller
{
    [self.inputController reset];
}

/**
 *  点击消息头像委托
 *  您可以通过该回调实现：跳转到对应用户的详细信息界面。
 */
- (void)messageController:(MSChatRoomMessageController *)controller onSelectMessageAvatar:(MSMessageCell *)cell
{
    if ([self.delegate respondsToSelector:@selector(chatController:onSelectMessageAvatar:)]) {
        [self.delegate chatController:self onSelectMessageAvatar:cell];
    }
    [self.inputController reset];
}

/**
 *  点击消息内容委托
 */
- (void)messageController:(MSChatRoomMessageController *)controller onSelectMessageContent:(MSMessageCell *)cell
{
    [self.inputController reset];
    if ([self.delegate respondsToSelector:@selector(chatController:onSelectMessageContent:)]) {
        [self.delegate chatController:self onSelectMessageContent:cell];
    }
}

/**
 *  显示长按菜单前的回调函数
 */
- (BOOL)messageController:(MSChatRoomMessageController *)controller willShowMenuInCell:(UIView *)view
{
    if ([self.inputController.inputBar.inputTextView isFirstResponder]) {
        self.inputController.inputBar.inputTextView.overrideNextResponder = view;
        return YES;
    }
    return NO;
}

- (void)didHideMenuInMessageController:(MSChatRoomMessageController *)controller
{
    self.inputController.inputBar.inputTextView.overrideNextResponder = nil;
}

- (void)sendMessage:(MSIMElem *)message
{
    [[MSIMManager sharedInstance] sendChatRoomMessage:message toRoomID:self.roomInfo.room_id successed:^(NSInteger msg_id) {
        
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
    if ([self.delegate respondsToSelector:@selector(chatController:didSendMessage:)]) {
        [self.delegate chatController:self didSendMessage:message];
    }
}
@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatRoom/MSChatRoomController.swift
//
//  MSChatRoomController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/11/2.
//

import UIKit
import MSIMSDK
import AVFoundation

public protocol MSChatRoomControllerDelegate: NSObjectProtocol {
    
    ///发送新消息时的回调
    func didSendMessage(controller: MSChatRoomController,elem: MSIMElem)
    ///每条新消息在进入气泡展示区之前，都会通知给您
    ///主要用于甄别自定义消息
    ///如果您返回 nil，MSChatViewController 会认为该条消息非自定义消息，会将其按照普通消息的处理流程进行处理。
    ///如果您返回一个 MSMessageCellData 类型的对象，MSChatViewController 会在随后触发的 onShowMessageData() 回调里传入您返回的 cellData 对象。
    ///也就是说，onNewMessage() 负责让您甄别自己的个性化消息，而 onShowMessageData() 回调则负责让您展示这条个性化消息。
    func prepareForMessage(controller: MSChatRoomController,elem: MSIMElem) -> MSMessageCellData?
    
    ///展示自定义个性化消息
    ///您可以通过重载 onShowMessageData() 改变消息气泡的默认展示逻辑，只需要返回一个自定义的 MSMessageCell 对象即可。
    func onShowMessageData(controller: MSChatRoomController,cellData: MSMessageCellData) -> MSMessageCell.Type?
    
    ///点击某一“更多”单元的回调委托
    func onSelectMoreCell(controller: MSChatRoomController,cell: MSInputMoreCell)
    
    ///点击消息头像回调
    func onSelectMessageAvatar(controller: MSChatRoomController,cell: MSMessageCell)
    
    ///点击消息内容回调
    func onSelectMessageContent(controller: MSChatRoomController,cell: MSMessageCell)
}

public class MSChatRoomController: UIViewController {

    
    public var room_id: String? {
        didSet {
            self.messageController.room_id = room_id
        }
    }
    
    public weak var delegate: MSChatRoomControllerDelegate?
    
    public private(set) var messageController: MSChatRoomMessageController = MSChatRoomMessageController()
    
    public private(set) var inputController: MSInputViewController!
    
    func sendMessage(message: MSIMElem) {
        if self.room_id == nil {
            MSHelper.showToastWithText(text: "room_id is nill")
            return
        }
        MSIMManager.sharedInstance().sendChatRoomMessage(message, toRoomID: self.room_id!) { _ in
            
        } failed: { _, desc in
            MSHelper.showToastFailWithText(text: desc ?? "")
        }
        delegate?.didSendMessage(controller: self, elem: message)
    }
    
    public override func viewDidLoad() {
        super.viewDidLoad()

        setupUI()
    }

    deinit {
        print("\(self) dealloc")
    }
    
    private func setupUI() {
        messageController.delegate = self
        messageController.room_id = self.room_id
        messageController.view.frame = CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.height - MSMcros.TTextView_Height - UIScreen.safeAreaBottomHeight)
        addChild(messageController)
        view.addSubview(messageController.view)
        
        inputController = MSInputViewController(chatType: .MSIM_CHAT_TYPE_CHATROOM, delegate: self)
        inputController.view.frame = CGRect(x: 0, y: UIScreen.height - MSMcros.TTextView_Height - UIScreen.safeAreaBottomHeight, width: UIScreen.width, height: MSMcros.TTextView_Height + UIScreen.safeAreaBottomHeight)
        inputController.view.autoresizingMask = .flexibleTopMargin
        addChild(inputController)
        view.addSubview(inputController.view)
    }
}

// MARK: - MSInputViewControllerDelegate
extension MSChatRoomController: MSInputViewControllerDelegate {
    public func didChangeHeight(inputController: MSInputViewController, height: CGFloat) {
        UIView.animate(withDuration: 0.3, delay: 0, options: .curveEaseOut) {
            
            var msgFrame = self.messageController.view.frame
            msgFrame.size.height = self.view.frame.size.height - height
            self.messageController.view.frame = msgFrame
            
            var inputFrame = self.inputController.view.frame
            inputFrame.origin.y = msgFrame.origin.y + msgFrame.size.height
            inputFrame.size.height = height
            self.inputController.view.frame = inputFrame
            self.messageController.scrollToBottom(animate: false)
            
        } completion: { _ in
            
        }
    }
    
    public func didSendTextMessage(inputController: MSInputViewController, msg: String) {
        let textElem = MSIMManager.sharedInstance().createTextMessage(msg)
        self.sendMessage(message: textElem)
    }
    
    public func didSendVoiceMessage(inputController: MSInputViewController, filePath: String) {
        
        let url = URL(fileURLWithPath: filePath)
        let audioAsset = AVURLAsset(url: url, options: nil)
        let duration = Int(CMTimeGetSeconds(audioAsset.duration))
        if let length = try? FileManager.default.attributesOfItem(atPath: filePath)[FileAttributeKey.size] as? Int {
            var voiceElem = MSIMVoiceElem()
            voiceElem.path = filePath
            voiceElem.duration = duration
            voiceElem.dataSize = length
            voiceElem = MSIMManager.sharedInstance().createVoiceMessage(voiceElem)
            self.sendMessage(message: voiceElem)
        }
    }
    
    public func contentDidChanged(inputController: MSInputViewController, text: String) {
        
    }
    
    public func inputControllerDidInputAt(inputController: MSInputViewController) {
        
    }
    
    public func didDeleteAt(inputController: MSInputViewController, atText: String) {
        
    }
    
    public func didSelectMoreCell(inputController: MSInputViewController, cell: MSInputMoreCell) {
        
        delegate?.onSelectMoreCell(controller: self, cell: cell)
    }
}

// MARK: MSChatRoomMessageControllerDelegate
extension MSChatRoomController: MSChatRoomMessageControllerDelegate {
    public func onRecieveSignalMessage(controller: MSChatRoomMessageController, elems: [MSIMElem]) {
        
    }
    
    public func prepareForMessage(controller: MSChatRoomMessageController, elem: MSIMElem) -> MSMessageCellData? {
        if let data = delegate?.prepareForMessage(controller: self, elem: elem) {
            return data
        }
        return nil
    }
    
    public func onShowMessageData(controller: MSChatRoomMessageController, cellData: MSMessageCellData) -> MSMessageCell.Type? {
        
        if let type = delegate?.onShowMessageData(controller: self, cellData: cellData) {
            return type
        }
        return nil
    }
    
    public func didTapInMessageController(controller: MSChatRoomMessageController) {
        self.inputController.reset()
    }
    
    public func onSelectMessageAvatar(controller: MSChatRoomMessageController, cell: MSMessageCell) {
        delegate?.onSelectMessageAvatar(controller: self, cell: cell)
        self.inputController.reset()
    }
    
    public func onSelectMessageContent(controller: MSChatRoomMessageController, cell: MSMessageCell) {
        delegate?.onSelectMessageContent(controller: self, cell: cell)
        self.inputController.reset()
    }
    
    public func willShowMenuInCell(controller: MSChatRoomMessageController, view: UIView) -> Bool {
        if self.inputController.inputBar.inputTextView.isFirstResponder {
            self.inputController.inputBar.inputTextView.overrideNextResponder = view
            return true
        }
        return false
    }
    
    public func didHideMenuInMessageController(controller: MSChatRoomMessageController) {
        self.inputController.inputBar.inputTextView.overrideNextResponder = nil
    }
}
